Random data

AnderGray opened the pull request:
Converting random nuclear data for use in OpenMC.

2 new files added

- convert_tendl_rand.py 
Downloads random tendl evaluations from PSI and converts them to hdf5. Only some nuclides are available from PSI, they're marked in the script along with their locations. So far this script knows about `O16, Si28, Si29, Si30, Fe54, Fe56, Fe57, Fe58`

- sample_sandy.py
Samples random (gaussian) evaluations of an endf file following its covariance matrix using SANDY, and converts it to hdf5. Sandy is required to be installed and in PATH


Both of these scripts were intended to create a `cross_sections.xml` file which copies the library from `OPENMC_CROSS_SECTIONS`, and adds random evaluations as eg. `Fe54-1`. The Monte Carlo propagation could then be done by changing the nuclide in the simulation from eg `Fe54` to `Fe54-1`, and iterating through all the random evaluations. 

However the creation of the `xml` file has been commented out at the end of both scripts, because it requires a change to `library.py` 's` register_file()` to accept nuclide names as an argument. Also some changes to `combine_libraries.py` would be needed.

AnderGray commented:
PR [#1580](https://github.com/openmc-dev/openmc/pull/1580) would allow for creation of the xml file.

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,360 @@
+    pass
+
+
+#n_choices=['all', 'O16', 'Na23', 'Si28', 'Si29','Si30', 'Fe054', 'Fe056', 'Fe057', 'Fe058', 'Ge','Pu240',]
```
Remove commented out code

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,360 @@
+
+
+#n_choices=['all', 'O16', 'Na23', 'Si28', 'Si29','Si30', 'Fe054', 'Fe056', 'Fe057', 'Fe058', 'Ge','Pu240',]
+n_choices=['all', 'O16','Si28', 'Si29','Si30', 'Fe54', 'Fe56', 'Fe57', 'Fe58']
```
Generally, we like to stay consistent with style by following the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guide. If you're too lazy to think about what the rules are, I'd recommend using a program like [black](https://pypi.org/project/black/) to autoformat your code.

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,360 @@
+
+args = parser.parse_args()
+
+# All availible online files from psi
+    # Name                                                      Size    Num
+#https://tendl.web.psi.ch/tendl_2017/tar_files/O016.random.tgz  60MB    641     x
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Na23.random.tgz  130MB   831     
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Si28.random.tgz  110MB   500     x
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Si29.random.tgz  100MB   500     x
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Si30.random.tgz  80MB    500     x
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Fe054.random.tgz 320MB   500     x
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Fe056.random.tgz 410MB   613     x
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Fe057.random.tgz 525MB   732     x
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Fe058.random.tgz 280MB   500     x
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Ge.random.tgz    1.3GB   ?   
+#https://tendl.web.psi.ch/tendl_2017/tar_files/Pu240.random.tgz 200MB   624 
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Al27.random.tgz  150MB   300     
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Si28.random.tgz  125MB   ...     x
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Ti48.random.tgz  140MB    ↓
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Ni58.random.tgz  210MB           x
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Cu63.random.tgz  205MB           
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Cu65.random.tgz  180MB
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Zr90.random.tgz  150MB
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Zr91.random.tgz  175MB
+#https://tendl.web.psi.ch/tendl_2015/tar_files/Zr92.random.tgz  170MB
+#https://tendl.web.psi.ch/tendl_2015/tar_files/W180.random.tgz  210MB
+#https://tendl.web.psi.ch/tendl_2015/tar_files/W182.random.tgz  215MB
+#https://tendl.web.psi.ch/tendl_2015/tar_files/W183.random.tgz  215MB
+#https://tendl.web.psi.ch/tendl_2015/tar_files/W184.random.tgz  250MB   ...
+#https://tendl.web.psi.ch/tendl_2015/tar_files/W186.random.tgz  180MB   300
+
+#RandomAce
+
+#https://tendl.web.psi.ch/tendl_2015/tar_files/H1.nuss.05.10.2016.tgz   11MB    340
+#https://tendl.web.psi.ch/tendl_2015/tar_files/O16.nuss.05.10.2016.tgz  570MB   340
+#https://tendl.web.psi.ch/tendl_2015/tar_files/U235.nuss.10.10.2016.tgz 430MB   100
+#https://tendl.web.psi.ch/tendl_2015/tar_files/U238.nuss.10.10.2016.tgz 1.5GB   100
+
+#ParsedNuclides._get_args('nuclides')
```
Remove this dead code

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,360 @@
+    outputDir = os.path.abspath(args.destination)
+
+
+endf_files_dir  = '/'.join([outputDir, 'endf'])
+ace_files_dir   = '/'.join([outputDir, 'ace'])
+hdf5_files_dir  = '/'.join([outputDir, 'hdf5'])
```
In other scripts, we've adopted use of the [pathlib](https://docs.python.org/3/library/pathlib.html) standard library module for handling file/directory names. I'd recommend going through your script and using pathlib throughout

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,360 @@
+        prefix = 'n-'
+        suffix = '-rand-'
+        for i in range(0,numFiles):
+            if i < 10:
+                OldNumber = '000' + str(i)
+            elif i < 100:
+                OldNumber = '00' + str(i)
+            elif i < 1000:
+                OldNumber = '0' + str(i)
```
`f'{i:04}'` would give a zero-padded string of length 4, achieving the same effect with one line of code

paulromano commented on sample_sandy.py:
```diff
@@ -0,0 +1,189 @@
+        sys.exit()
+    atomicNum = atomicDict[atomicSym]
+
+    if massNum < 10:
+        fileMass = '00' + str(massNum)
+    elif massNum < 100:
+        fileMass = '0' + str(massNum)
+
+    if atomicNum < 10:
+        fileAtomic = '00' + str(atomicNum)
+    elif atomicNum < 100:
+        fileAtomic = '0' + str(atomicNum)
```
Same suggestion about zero padding

paulromano commented on sample_sandy.py:
```diff
@@ -0,0 +1,189 @@
+    if not os.path.exists(nucDirEndf):
+        os.mkdir(nucDirEndf)
+    shutil.copyfile(libdir+'/'+'neutron'+'/'+nucDict[nuc]['fileName'], nucDirEndf  + '/'+nucDict[nuc]['fileName'])
+    #os.chdir(nucDirEndf)
+    #subprocess.call("cp {} .".format(libdir+'/'+nucDict[nuc]['fileName']))
```
Remove

AnderGray commented:
Hi @paulromano, thanks for the review. Sorry I took so long getting back to you. 

I think I've addressed all of your suggestions. I've also added the other compatible nuclides from the random evaluations from TENDL. They have more available, but the file format is different for older years, so I've only included the 2017 evaluations which have the same format.

There is also a format only option in both scripts, for when you don't want to sample/download the endf files.

The scripts should also now work without the need for PR [#1580](https://github.com/openmc-dev/openmc/pull/1580) or changes to `combine_libraries.py` .

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+        uncompressedFileSize += nuclide_details[i]["fileSize"]
+        NumOfFiles += nuclide_details[i]["fileNum"]
+
+    downloadSize = "{} MB".format(downloadFileSize)
+    uncomFileSize = "{} MB".format(uncompressedFileSize)
```
```suggestion
    downloadSize = f"{downloadFileSize} MB"
    uncomFileSize = f"{uncompressedFileSize} MB"
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+    downloadSize = "{} MB".format(downloadFileSize)
+    uncomFileSize = "{} MB".format(uncompressedFileSize)
+    if downloadFileSize > 1000:
+        downloadSize = "{} GB".format(downloadFileSize / 1000)
```
```suggestion
        downloadSize = f"{downloadFileSize / 1000} GB"
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+    if downloadFileSize > 1000:
+        downloadSize = "{} GB".format(downloadFileSize / 1000)
+    if uncompressedFileSize > 1000:
+        uncomFileSize = "{} GB".format(uncompressedFileSize / 1000)
```
```suggestion
        uncomFileSize = f"{uncompressedFileSize / 1000} GB"
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+            + nuclide_details[nucs]["webname"]
+            + release_details[nuclide_details[nucs]["release"]]["ending"]
+        )
+        print("Downloading {}...".format(nucs))
```
```suggestion
        print(f"Downloading {nucs}...")
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+
+    Are you sure you want to continue? ([y]/n)
+    """.format(
+        downloadSize, uncomFileSize, NumOfFiles, list_
```
PEP8: For local variables, you should generally be using `names_with_underscores` rather than camelCase

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+            prefix = "n-"
+            suffix = "-rand-"
+
+            for i in range(0, numFiles):
```
```suggestion
            for i in range(numFiles):
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+            for i in range(0, numFiles):
+                OldNumber = f"{i:04}"
+                OldFile = prefix + nuclide_details[nucs]["filename"] + suffix + OldNumber
+                newFile = nucs + "-" + str(i + 1)
```
```suggestion
                newFile = f"{nucs}-{i+1}"
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+                newFile = nucs + "-" + str(i + 1)
+                
+                if nuclide_details[nucs]["gunzip"]:
+                    os.system("gunzip " + str(outDir) + OldFile + ".gz")
```
```suggestion
                    os.system(f"gunzip {outDir} {OldFile}.gz")
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+                
+                if nuclide_details[nucs]["gunzip"]:
+                    os.system("gunzip " + str(outDir) + OldFile + ".gz")
+                os.rename(os.path.join(outDir, OldFile), os.path.join(outDir, newFile))
```
```suggestion
                (outDir / OldFile).rename(outDir / newFile)
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+    """Process ENDF neutron sublibrary file into HDF5 and write into a
+    specified output directory."""
+
+    fileIn = inDir / (nuc + "-" + str(i))
+    fileOut = outDir / (nuc + "-" + str(i) + ".h5")
```
```suggestion
    fileIn = inDir / f"{nuc}-{i}"
    fileOut = outDir / f"{nuc}-{i}.h5"
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+    fileIn = inDir / (nuc + "-" + str(i))
+    fileOut = outDir / (nuc + "-" + str(i) + ".h5")
+
+    data = openmc.data.IncidentNeutron.from_njoy(fileIn)  # , temperatures=293.6)
```
```suggestion
    data = openmc.data.IncidentNeutron.from_njoy(fileIn)
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+    fileOut = outDir / (nuc + "-" + str(i) + ".h5")
+
+    data = openmc.data.IncidentNeutron.from_njoy(fileIn)  # , temperatures=293.6)
+    data.name = nuc + "-" + str(i)
```
```suggestion
    data.name = f"{nuc}-{i}"
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+    data.name = nuc + "-" + str(i)
+    data.export_to_hdf5(fileOut, "w")
+    if i % 40 == 0:
+        print("Nuclide " + nuc + " " + str(i) + "/" + str(fileNum) + " finished")
```
```suggestion
        print(f"Nuclide {nuc} {i}/{fileNum} finished")
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+        print("Nuclide " + nuc + " " + str(i) + "/" + str(fileNum) + " finished")
+
+
+print("Beginning njoy processing")
```
```suggestion
print("Beginning NJOY processing")
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+        hdf5_files_dir.mkdir(exist_ok=True)
+        outDir.mkdir(exist_ok=True)
+
+        print("Beginning nuclide " + nuc + " ...")
```
```suggestion
        print(f"Beginning nuclide {nuc} ...")
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+    outDir = hdf5_files_dir / nuc
+    for i in range(1, fileNum + 1):
+        fileOut = outDir / (nuc + "-" + str(i) + ".h5")
+        #lib.register_file(fileOut, nuc + "-" + str(i))
+        lib.register_file(fileOut)
```
```suggestion
        lib.register_file(fileOut)
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+    fileNum = nuclide_details[nuc]["fileNum"]
+    outDir = hdf5_files_dir / nuc
+    for i in range(1, fileNum + 1):
+        fileOut = outDir / (nuc + "-" + str(i) + ".h5")
```
```suggestion
        fileOut = outDir / f"{nuc}-{i}.h5"
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+else:
+    lib.export_to_xml(post)
+
+os.remove(pre)
```
```suggestion
pre.unlink()
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+post = outputDir / "cross_sections_Tendl.xml"
+
+lib.export_to_xml(pre)
+if os.path.exists(post):
```
```suggestion
if post.exists():
```

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+if args.destination is None:
+    outputDir = scriptDir / library_name
+else:
+    outputDir = Path(os.path.abspath(args.destination))
```
```suggestion
    outputDir = Path(args.destination).resolve()
```

AnderGray commented:
Hi @paulromano, I think I've incorporated all your suggestions into `sample_sandy.py`. There may still be some camelCase in there. I'm not so familiar with PEP8. Do you want all local variables with underscore? eg. `fileNum`, `inDir` ,`OldNumber`?

Cheers, A

AnderGray commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,361 @@
+                    help="Directory to create new library in")
+parser.add_argument("-b", "--batch", action="store_true", 
+                    help="supresses standard in")
+parser.add_argument("-f", "--formatOnly", default=False,
```
I take it you meant `format_only`. [Commit 36](https://github.com/openmc-dev/data/pull/36/commits/978f4a08dd81c3c6b1d9e8fdeb585b2cd7ea9cc2)

paulromano commented:
Thanks for the updates @AnderGray. Yes, local variables should be lowercase with underscores, i.e. `file_num` `in_dir` and `old_number`

AnderGray commented:
Thanks @paulromano, I think I've caught everything

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,356 @@
+    description=description, formatter_class=CustomFormatter
+)
+parser.add_argument("-n", "--nuclides", choices=n_choices, nargs="+",
+                    default="Fe56", help="The nuclides to be downloaded. Available are: "
```
```suggestion
                    default=["Fe56"], help="The nuclides to be downloaded. Available are: "
```
Without this, the script currently fails if you let it use the default

paulromano commented on convert_tendl_rand.py:
```diff
@@ -0,0 +1,356 @@
+    download_warning = """
+    WARNING: This script will download {} of 
+    data, which is {} of data when processed. 
+    This corresponds to {} random crossections.
```
```suggestion
    This corresponds to {} random cross sections.
```

paulromano commented on sample_sandy.py:
```diff
@@ -0,0 +1,183 @@
+    file_name = f"{prefix}{file_atomic}_{atomic_sym}_{file_mass}{suffix}"
+
+    if not (libdir / "neutron" / file_name).isfile():
+        print(f"File {libdir / "neutron" / file_name} does not exist")
```
```suggestion
        print(f"File {libdir / 'neutron' / file_name} does not exist")
```

paulromano commented on sample_sandy.py:
```diff
@@ -0,0 +1,183 @@
+            nuc_dir_endf / nuc_dict[nuc]["file_name"],
+        )
+        os.chdir(nuc_dir_endf)
+        sandy_command = f"sandy {nuc_dict[nuc]["file_name"]} --samples {args.samples} --outname {nuc} --processes {args.processes}"
```
```suggestion
        sandy_command = f"sandy {nuc_dict[nuc]['file_name']} --samples {args.samples} --outname {nuc} --processes {args.processes}"
```

paulromano commented on sample_sandy.py:
```diff
@@ -0,0 +1,183 @@
+for nuc in nuclides:
+    out_dir = hdf5_files_dir / nuc
+    for i in range(1, file_num + 1):
+        fileOut = out_dir / f"{nuc}-{i}.h5")
```
```suggestion
        fileOut = out_dir / f"{nuc}-{i}.h5"
```

paulromano commented on sample_sandy.py:
```diff
@@ -0,0 +1,183 @@
+endf_files_dir = output_dir / "endf"
+hdf5_files_dir = output_dir / "hdf5"
+
+libdir = args.libdir
+if libdir == None:
+    libdir = Path(os.getenv("NUCLEAR_DATA_DIR")) / "nndc-b7.1-endf"
```
Personally, I would just make libdir a required argument rather than an optional argument that uses this (undocumented) environment variable as a fallback

AnderGray commented:
Hi @paulromano, sorry for taking so long to get back to you. How do the last few commits look?

paulromano commented:
Thanks for the contribution @AnderGray!
